extract_embeddings.py

```diff
@@ -48,13 +68,123 @@
     id=int,
     document_id=int,
     text=str,
+    header=str,
     embedding=bytes,
     created_at=datetime,
-    pk="id",
-    foreign_keys=[("document_id", "documents")],
+    pk="id",    foreign_keys=[("document_id", "documents")],
     ignore=True,
 )
 
+# -----------------------------------------------------
+# Header pattern (Markdown)
+# -----------------------------------------------------
+HEADING_PATTERN = re.compile(r'^(#{1,6})\s+(.*)$')
+
+# -----------------------------------------------------
+# Header processing functions
+# -----------------------------------------------------
+def get_heading_level(line: str):
+    """
+    Returns the heading level and text,
+    or (None, None) if the line is not a heading.
+    """
+    match = HEADING_PATTERN.match(line)
+    if match:
+        hashes = match.group(1)
+        heading_text = match.group(2).strip()
+        level = len(hashes)
+        return level, heading_text
+    return None, None
+
+
+def update_open_headings(open_headings, line):
+    """
+    Updates the list of open headings according to the line.
+    
+    - If an H1 is found, the list is reset.
+    - If the line is a heading of level >1:
+        * If the list is empty, it is added.
+        * If the last open heading has a lower or equal level,
+          it is added without removing the previous one (to preserve siblings).
+        * If the new heading is of a higher level (lower number)
+          than the last one, those with a level lower than the new one
+          are preserved and the heading is added.
+    """
+    lvl, txt = get_heading_level(line)
+    if lvl is None:
+        # Line without heading, state is not modified
+        return open_headings
+
+    if lvl == 1:
+        # An H1 closes all previous context.
+        return [(1, txt)]
+    else:
+        if not open_headings:
+            return [(lvl, txt)]
+        else:
+            # If the last heading has a lower or equal level, the current one is added.
+            if open_headings[-1][0] <= lvl:
+                open_headings.append((lvl, txt))
+            else:
+                # If the new heading is of a higher level (more important)
+                # only headings that are of a lower level (higher) than the new one are preserved.
+                new_chain = [item for item in open_headings if item[0] < lvl]
+                new_chain.append((lvl, txt))
+                open_headings = new_chain
+        return open_headings
+
+
+def build_header(doc_title: str, page: str, open_headings: list, chunk_number: int):
```

Comment:
if you go with the above comment, this function would change to something like:

```python
def build_header_dict(doc_title: str, page: str, open_headings: dict[int, str], chunk_number: int):
    header_lines = [f"# Document: {doc_title} | page: {page}"]
    
    for level in sorted(open_headings.keys()):
        text = open_headings[level]
        header_lines.append(f"{'#' * level} {text}")
    
    return "\n".join(header_lines)
```

By the way, I don't think we need a special case for when `chunk_humber == 1`.